firedrake/assemble_expressions.py

```diff
@@ -1,645 +1,424 @@
+from ufl.algorithms.apply_algebra_lowering import LowerCompoundAlgebra
+from ufl.classes import Index, MultiIndex, ConstantValue
+from ufl.corealg.map_dag import map_expr_dags
+from ufl.corealg.multifunction import MultiFunction
+from ufl.corealg.traversal import unique_pre_traversal as ufl_traversal
 
-from firedrake import constant
-from firedrake import function
-from firedrake import utils
+import firedrake
+from firedrake.utils import ScalarType, cached_property, known_pyop2_safe
 
-from functools import singledispatch
 
+def extract_coefficients(expr):
+    return tuple(e for e in ufl_traversal(expr) if isinstance(e, ufl.Coefficient))
 
-def ufl_type(*args, **kwargs):
-    r"""Decorator mimicing :func:`ufl.core.ufl_type.ufl_type`.
 
-    Additionally adds the class decorated to the appropriate set of ufl classes."""
-    def decorator(cls):
-        orig_ufl_type(*args, **kwargs)(cls)
-        classes.all_ufl_classes.add(cls)
-        if cls._ufl_is_abstract_:
-            classes.abstract_classes.add(cls)
-        else:
-            classes.ufl_classes.add(cls)
-        if cls._ufl_is_terminal_:
-            classes.terminal_classes.add(cls)
+class Translator(MultiFunction, ufl2gem.Mixin):
+    def __init__(self):
+        self.varmapping = OrderedDict()
+        MultiFunction.__init__(self)
+        ufl2gem.Mixin.__init__(self)
+
+    # Override shape-based things
+    # Need to inspect GEM shape not UFL shape, due to Coefficients changing shape.
+    def sum(self, o, *ops):
+        shape, = set(o.shape for o in ops)
+        indices = gem.indices(len(shape))
+        return gem.ComponentTensor(gem.Sum(*[gem.Indexed(op, indices) for op in ops]),
+                                   indices)
+
+    def real(self, o, expr):
+        indices = gem.indices(len(expr.shape))
+        return gem.ComponentTensor(gem.MathFunction('real', gem.Indexed(expr, indices)),
+                                   indices)
+
+    def imag(self, o, expr):
+        indices = gem.indices(len(expr.shape))
+        return gem.ComponentTensor(gem.MathFunction('imag', gem.Indexed(expr, indices)),
+                                   indices)
+
+    def conj(self, o, expr):
+        indices = gem.indices(len(expr.shape))
+        return gem.ComponentTensor(gem.MathFunction('conj', gem.Indexed(expr, indices)),
+                                   indices)
+
+    def abs(self, o, expr):
+        indices = gem.indices(len(expr.shape))
+        return gem.ComponentTensor(gem.MathFunction('abs', gem.Indexed(expr, indices)),
+                                   indices)
+
+    def conditional(self, o, condition, then, else_):
+        assert condition.shape == ()
+        shape, = set([then.shape, else_.shape])
+        indices = gem.indices(len(shape))
+        return gem.ComponentTensor(gem.Conditional(condition, gem.Indexed(then, indices),
+                                                   gem.Indexed(else_, indices)),
+                                   indices)
```

Comment:
Kind of late comment, but this pattern of `gem.ComponentTensor(Foo(gem.Indexed(bar, indices), ...), indices)` comes up so often that it would be useful to capture it as helper function.  The helper function could probably go to `ufl2gem.Mixin` (and also used there) and then imported here.